### PR TITLE
Delete my User Account functionality

### LINKED ISSUE #404: 🌟 [EPIC] User Settings
> [ulfgebhardt](https://github.com/ulfgebhardt) **Authored by [ulfgebhardt](https://github.com/ulfgebhardt)**
_<time datetime="2019-04-04T14:57:29Z" title="Thursday, April 4th 2019, 4:57:29 pm +02:00">Apr 4, 2019</time>_
_Closed <time datetime="2019-12-19T21:14:27Z" title="Thursday, December 19th 2019, 10:14:27 pm +01:00">Dec 19, 2019</time>_
---

### User Settings
image

- [X] Header "Settings"
- [ ] Navigation Crumbs "Settings / Subpage" https://github.com/Human-Connection/Human-Connection/issues/405
- [X] Save Settings (on each Page)

Your Data:
- [X] Name
- [x] Name changeable - Bug https://github.com/Human-Connection/Human-Connection/issues/376
- [ ] EMail https://github.com/Human-Connection/Human-Connection/issues/407
- [ ] Avatar https://github.com/Human-Connection/Human-Connection/issues/408

Security:
- [X] Current password
- [x] New Password -> Bug https://github.com/human-connection/human-connection/issues/247
- [X] New Password Repeat
- [x] Password strength https://github.com/Human-Connection/Human-Connection/issues/410

(optional) Invites:
- [ ] Generate invite Code Button
- [ ] Display Invite Codes + Copy Link Button
- [ ] Invite System

(later) Download Data:
- [ ] Send Download Link Button + Functionality https://github.com/Human-Connection/Human-Connection/issues/211

Delete Account https://github.com/Human-Connection/Human-Connection/issues/404:
- [x] Confirm slider https://github.com/Human-Connection/Human-Connection/issues/404
- [x] Delete Button + Functionality https://github.com/Human-Connection/Human-Connection/issues/404

(optional) My Organizations:
- [ ] Create New Orga Button
- [ ] Select existing Orgas -> Profile
- [ ] Delete existing Orga
- [ ] Existing Orga Settings

Languages https://github.com/Human-Connection/Human-Connection/issues/409:
- [ ] Content Languages enable/disable https://github.com/Human-Connection/Human-Connection/issues/409
- [ ] Interface Language https://github.com/Human-Connection/Human-Connection/issues/409

(optional) Presentation:
- [ ] Hide Users slider + Functionality

### Design & Layout - Alpha

image
image
image
image
image
image
image
image
image
image

### Design & Layout - Nitro

image
image
image
image
image
image
image

### PR DESCRIPTION
**Pull request by Tirokk**
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-05T00:24:53Z" title="Wednesday, June 5th 2019, 2:24:53 am +02:00">Jun 5, 2019</time>_
_Merged <time datetime="2019-06-14T16:45:31Z" title="Friday, June 14th 2019, 6:45:31 pm +02:00">Jun 14, 2019</time>_
---

### Issues

- fixes #404

### CODE REVIEW

**Comment by Tirokk** [comment]
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-05T00:26:17Z" title="Wednesday, June 5th 2019, 2:26:17 am +02:00">Jun 5, 2019</time>_

---


progress so far
Screenshot at 2019-06-04 21:22:57

we don't have any checkmarks as part of our styleguide :woman_shrugging: so I have resorted to a `select`, but open to other ideas... maybe a modal to ask for confirmation(?)

**Comment by Tirokk** [comment]
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-07T21:46:07Z" title="Friday, June 7th 2019, 11:46:07 pm +02:00">Jun 7, 2019</time>_

---


Here are some gifs with the functionality:
Delete user, but not posts or comments:
2019-06-07-delete-user-account-not-comments-or-posts
Delete user, with posts, but not comments:
2019-06-07-delete-user-account-and-posts
Delete user, with posts and comments:
2019-06-07-delete-user-account-posts-and-comments
I changed the styling slightly, which can be seen in the screenshots :point_down: 
Screenshot at 2019-06-07 18:38:08

Mobile Responsiveness:
Screenshot at 2019-06-07 18:40:07
Screenshot at 2019-06-07 18:40:37
Screenshot at 2019-06-07 18:38:35

File: backend/src/middleware/permissionsMiddleware.js [hunk 1]
```diff
@@ -93,6 +93,12 @@ const isAuthor = rule({
   return authorId === user.id
 })
 
+const isDeletingOwnAccount = rule({
```
Comment (Tirokk) [inline backend/src/middleware/permissionsMiddleware.js:96]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-07T22:17:01Z" title="Saturday, June 8th 2019, 12:17:01 am +02:00">Jun 8, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -80,6 +80,12 @@ const isAuthor = rule({
   return authorId === user.id
 })
 
+const isDeletingOwnAccount = rule({
```


Oh yes, this is quite important!

File: backend/src/schema/resolvers/users.js [hunk 2]
```diff
@@ -11,5 +11,27 @@ export default {
...
+    DeleteUser: async (object, params, context, resolveInfo) => {
+      const { resource } = params
+      const session = context.driver.session()
+
+      if (resource && resource.length) {
+        await Promise.all(
+          resource.map(async node => {
+            await session.run(
+              `
+            MATCH (resource:${node})<-[:WROTE]-(author:User {id: $userId})
```
Comment (Tirokk) [inline backend/src/schema/resolvers/users.js:23]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-07T22:17:19Z" title="Saturday, June 8th 2019, 12:17:19 am +02:00">Jun 8, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -11,5 +11,27 @@ export default {
       params = await fileUpload(params, { file: 'avatarUpload', url: 'avatar' })
       return neo4jgraphql(object, params, context, resolveInfo, false)
     },
+    DeleteUser: async (object, params, context, resolveInfo) => {
+      const { resource } = params
+      const session = context.driver.session()
+
+      if (resource && resource.length) {
+        await Promise.all(
+          resource.map(async node => {
+            await session.run(
+              `
+            MATCH (resource:${node})<-[:WROTE]-(author:User {id: $userId})
```


 A couple of thoughts here: First of all, I believe, you can do this in one statement. I would try sth. like:
```
MATCH (author:User {id: $userId}), (post:Post)<-[:WROTE]-(author), (comment:Comment)<[:WROTE]-(author)
DETACH DELETE author, post, comment
```
I don't know if it works like that but I guess you get the point.

File: backend/src/schema/resolvers/users.js [hunk 3]
```diff
@@ -11,5 +11,27 @@ export default {
...
+            SET resource.deleted = true
+            RETURN author`,
+              {
+                userId: context.user.id,
+              },
+            )
+          }),
+        )
+        session.close()
```
Comment (Tirokk) [inline backend/src/schema/resolvers/users.js:32]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-07T22:21:56Z" title="Saturday, June 8th 2019, 12:21:56 am +02:00">Jun 8, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -11,5 +11,27 @@ export default {
       params = await fileUpload(params, { file: 'avatarUpload', url: 'avatar' })
       return neo4jgraphql(object, params, context, resolveInfo, false)
     },
+    DeleteUser: async (object, params, context, resolveInfo) => {
+      const { resource } = params
+      const session = context.driver.session()
+
+      if (resource && resource.length) {
+        await Promise.all(
+          resource.map(async node => {
+            await session.run(
+              `
+            MATCH (resource:${node})<-[:WROTE]-(author:User {id: $userId})
+            DETACH DELETE resource
+            RETURN author`,
+              {
+                userId: context.user.id,
+              },
+            )
+          }),
+        )
+        session.close()
```


Most importantly is that if you delete the nodes, let's say you delete a post, then you won't delete all the comments on that post, because they belong to a different user. So my suggestion is not to delete a post but delete **all attributes** of that post.

I would say it's good practice to have something like a tombstone to maintain the context and the associations with the data that is already there. This is even more important with a comment. If you have a conversation with deleted comments and you cannot see anymore where a comment has been deleted, the responses to that comment might look awkward and out of context.

File: webapp/locales/en.json [hunk 4]
```diff
@@ -82,8 +82,14 @@
     "download": {
       "name": "Download Data"
     },
-    "delete": {
-      "name": "Delete Account"
+    "deleteUserAccount": {
+      "name": "Delete Data",
+      "contributionsCount": "Delete my {count} posts",
+      "commentsCount": "Delete my {count} comments",
+      "accountDescription": "Be aware that your Post and Comments are important to our community. If you still choose to delete them, you have to mark them below.",
```
Comment (Tirokk) [inline webapp/locales/en.json:89]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-07T22:27:55Z" title="Saturday, June 8th 2019, 12:27:55 am +02:00">Jun 8, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -74,7 +74,12 @@
       "name": "Download Data"
     },
     "delete": {
-      "name": "Delete Account"
+      "name": "Delete my User Account",
+      "countPosts": "Delete my {count} posts",
+      "countComments": "Delete my {count} comments",
+      "accountDescription": "Be aware that your Post and Comments are important to our community. If you still choose to delete those you have to mark them below.",
+      "accountWarning": "You <b>CAN'T MANAGE</b> and <b>CAN'T RECOVER</b> your Account, Posts... after deleting your account!",
+      "success": "Account successfully deleted"
```


Use https://www.deepl.com/translator for German translations and you will get an extra translation check from me for free :wink:

File: backend/src/schema/resolvers/users.js (hunk 2 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/users.js:23]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-07T22:39:44Z" title="Saturday, June 8th 2019, 12:39:44 am +02:00">Jun 8, 2019</time>_

---


the point is they wouldn't have the choice of deleting just their account, or just their account and posts, just their account and comments, etc...

Comment (Tirokk) [reply backend/src/schema/resolvers/users.js:23]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-07T22:47:53Z" title="Saturday, June 8th 2019, 12:47:53 am +02:00">Jun 8, 2019</time>_

---


Oh, right, true! We cannot customize `neo4j-graphql-js`.

**Comment by Tirokk** [comment]
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-10T17:10:27Z" title="Monday, June 10th 2019, 7:10:27 pm +02:00">Jun 10, 2019</time>_

---


I have changed up the implementation, and followed some other of your ideas @roschaefer... here is the new functionality

Deleting only my account:
2019-06-10 -delete-user-account-not-comments-or-posts

Deleting my account, and posts:
2019-06-10-delete-user-account-and-posts

Deleting my account, and comments:
2019-06-10-delete-user-account-and-comments

Deleting my account, posts, and comments:
2019-06-10 -delete-user-account-posts-and-comments

New look,
Screenshot at 2019-06-10 14:47:26

In German,
Screenshot at 2019-06-10 14:46:52
Screenshot at 2019-06-10 14:47:09

Mobile Responsiveness:
Screenshot at 2019-06-10 14:50:39
Screenshot at 2019-06-10 14:50:49
Screenshot at 2019-06-10 14:50:21
Screenshot at 2019-06-10 14:50:08

File: backend/src/schema/resolvers/users.js (hunk 3 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/users.js:32]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-10T17:34:15Z" title="Monday, June 10th 2019, 7:34:15 pm +02:00">Jun 10, 2019</time>_

---


hey @roschaefer, I think I followed all of your advice, or at least tried to...
I agree that it would be good to have some "tombstone", if a user deletes their comment, either by deleting their account, with all its comments, or by deleting a single comment.
It feels to me that this functionality is a separate ticket though, don't you agree? 
The reason I feel like that is cause the simple action of deleting one comment doesn't support it, much less deleting all comments.

Comment (Tirokk) [reply backend/src/schema/resolvers/users.js:32]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-10T17:37:26Z" title="Monday, June 10th 2019, 7:37:26 pm +02:00">Jun 10, 2019</time>_

---


Maybe when a user deletes a comment they have made, instead of deleting the node, we could delete all attributes, except the `id`, and change the content to `this comment has been deleted by the user`, plus add a `deletedOn` date?

**Comment by Tirokk** [comment]
> [ulfgebhardt](https://github.com/ulfgebhardt) **Authored by [ulfgebhardt](https://github.com/ulfgebhardt)**
_<time datetime="2019-06-11T12:31:32Z" title="Tuesday, June 11th 2019, 2:31:32 pm +02:00">Jun 11, 2019</time>_

---


@mattwr18 Can we name it "Delete Account" for a shorter name in the menu? <3

File: backend/src/schema/resolvers/users.js (hunk 3 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/users.js:32]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-11T14:35:45Z" title="Tuesday, June 11th 2019, 4:35:45 pm +02:00">Jun 11, 2019</time>_

---


Yes, that is the implementation of the alpha! We currently save a boolean flag on nodes `deleted = true`. Setting a date would be better, but we have to change the implementation of all the counters etc. I would suggest to open another issue for refactoring the `deleted` and `disabled` flags from type Boolean to type DateTime.

Comment (Tirokk) [reply backend/src/schema/resolvers/users.js:32]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-11T19:05:54Z" title="Tuesday, June 11th 2019, 9:05:54 pm +02:00">Jun 11, 2019</time>_

---


So, you suggesting that instead of deleting the comment, we should be just changing the attribute `delete = true`?

And for `posts`, you would expect that we not delete them completely because there are other user's comments on them? just delete the user relationship and it's content??

**Comment by Tirokk** [comment]
> [appinteractive](https://github.com/appinteractive) **Authored by [appinteractive](https://github.com/appinteractive)**
_<time datetime="2019-06-11T20:31:05Z" title="Tuesday, June 11th 2019, 10:31:05 pm +02:00">Jun 11, 2019</time>_

---


Hey guys, nice progress! Just the thing that you want the user to type that for every option is a bit of overkill from my perspective! Also does it not encourage anymore to leave the content in my opinion. I would advocate for using the checkboxes for the option again.

File: backend/src/schema/resolvers/users.js (hunk 3 above)
Comment (Tirokk) [reply backend/src/schema/resolvers/users.js:32]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-11T21:41:34Z" title="Tuesday, June 11th 2019, 11:41:34 pm +02:00">Jun 11, 2019</time>_

---


Regarding the comments, yes. Regarding the posts: Up to you. You won't see deleted posts in your newsfeed. So basically removing the posts will certainly make the posts along with the comments unlocatable. There will be probably no difference from the end user perspective, however we administrators could make use of the data if the relationships are still there at least.

Comment (Tirokk) [reply backend/src/schema/resolvers/users.js:32]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-13T00:40:10Z" title="Thursday, June 13th 2019, 2:40:10 am +02:00">Jun 13, 2019</time>_

---


taken care of in the latest commit

**Comment by Tirokk** [comment]
> [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-13T00:51:05Z" title="Thursday, June 13th 2019, 2:51:05 am +02:00">Jun 13, 2019</time>_

---





Screenshot at 2019-06-12 21:48:42
Screenshot at 2019-06-12 21:48:24
Screenshot at 2019-06-12 21:49:02
Screenshot at 2019-06-12 21:49:47
Screenshot at 2019-06-12 21:49:57
Screenshot at 2019-06-13 07:35:43

File: webapp/locales/de.json [hunk 5]
```diff
@@ -82,8 +82,14 @@
     "download": {
       "name": "Daten herunterladen"
     },
-    "delete": {
-      "name": "Konto löschen"
```
Comment (Tirokk) [inline webapp/locales/de.json:86]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-13T17:48:13Z" title="Thursday, June 13th 2019, 7:48:13 pm +02:00">Jun 13, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -25,16 +25,7 @@
     "shouted": "Empfohlen",
     "commented": "Kommentiert",
     "userAnonym": "Anonymus",
-    "socialMedia": "Wo sonst finde ich",
-    "network": {
```


@mattwr18 why did you delete the translations? :thinking:

File: webapp/pages/profile/_id/_slug.vue [hunk 6]
```diff
@@ -228,7 +228,6 @@
 
 <script>
 import uniqBy from 'lodash/uniqBy'
-
 import User from '~/components/User'
```
Comment (Tirokk) [inline webapp/pages/profile/_id/_slug.vue:231]: > [roschaefer](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-06-13T17:48:34Z" title="Thursday, June 13th 2019, 7:48:34 pm +02:00">Jun 13, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -69,13 +69,11 @@
         </ds-card>
         <ds-space />
         <ds-heading tag="h3" soft style="text-align: center; margin-bottom: 10px;">
-          {{ $t('profile.network.title') }}
+          Netzwerk
```


Please don't untranslate :heart:

File: webapp/locales/de.json (hunk 5 above)
Comment (Tirokk) [reply webapp/locales/de.json:86]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-13T18:18:44Z" title="Thursday, June 13th 2019, 8:18:44 pm +02:00">Jun 13, 2019</time>_

---


I didn't! well, I must have worked through a merge conflict incorrectly... I'll add it back
weird thing is that trying to pull in the latest from master says it's up to date :woman_shrugging:

File: webapp/pages/profile/_id/_slug.vue (hunk 6 above)
Comment (Tirokk) [reply webapp/pages/profile/_id/_slug.vue:231]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-06-13T18:19:07Z" title="Thursday, June 13th 2019, 8:19:07 pm +02:00">Jun 13, 2019</time>_

---


Also, don't know what happened here
:woman_shrugging:

**Comment by Tirokk** [comment]
> [codecov[bot]](https://github.com/apps/codecov) **Authored by [codecov[bot]](https://github.com/apps/codecov)**
_<time datetime="2019-06-14T12:42:49Z" title="Friday, June 14th 2019, 2:42:49 pm +02:00">Jun 14, 2019</time>_

---


# [Codecov](https://codecov.io/gh/Human-Connection/Human-Connection/pull/740?src=pr&el=h1) Report
> :exclamation: No coverage uploaded for pull request base (`master@d12d747`). [Click here to learn what that means](https://docs.codecov.io/docs/error-reference#section-missing-base-commit).
> The diff coverage is `0%`.

[Impacted file tree graph](https://codecov.io/gh/Human-Connection/Human-Connection/pull/740?src=pr&el=tree)

```diff
@@            Coverage Diff            @@
##             master     #740   +/-   ##
=========================================
  Coverage          ?   23.83%           
=========================================
  Files             ?      166           
  Lines             ?     2337           
  Branches          ?      123           
=========================================
  Hits              ?      557           
  Misses            ?     1745           
  Partials          ?       35
```


| [Impacted Files](https://codecov.io/gh/Human-Connection/Human-Connection/pull/740?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [webapp/pages/settings.vue](https://codecov.io/gh/Human-Connection/Human-Connection/pull/740/diff?src=pr&el=tree#diff-d2ViYXBwL3BhZ2VzL3NldHRpbmdzLnZ1ZQ==) | `0% <ø> (ø)` | |
| [webapp/pages/settings/delete-account.vue](https://codecov.io/gh/Human-Connection/Human-Connection/pull/740/diff?src=pr&el=tree#diff-d2ViYXBwL3BhZ2VzL3NldHRpbmdzL2RlbGV0ZS1hY2NvdW50LnZ1ZQ==) | `0% <0%> (ø)` | |
| [webapp/components/DeleteAccount/DeleteAccount.vue](https://codecov.io/gh/Human-Connection/Human-Connection/pull/740/diff?src=pr&el=tree#diff-d2ViYXBwL2NvbXBvbmVudHMvRGVsZXRlQWNjb3VudC9EZWxldGVBY2NvdW50LnZ1ZQ==) | `0% <0%> (ø)` | |

**Review by Mogge — CHANGES_REQUESTED** [review]
Ok, this is it from my side so far. :slightly_smiling_face:

Regarding your question @mattwr18 I would do it in the same way as other apps do it for destructive operations: Ask the user to type in the name of the resource that should be destroyed. So, if the user asks for a self-destruction, you could ask the user to type in the user's name.

**Review by Mogge — APPROVED** [review]
You did an astounding work @mattwr18 !!!
This is really a master piece 😍

Thanks for that, my friend !!! 😘🚀

Dog Clapping

I have pushed up some little german translation approvements …
Hope that’s okay …

The following will not keeping me from approving !!!

Should the user not only type their name, but their login data, to delete his/her account?
May you make an issue, if you agree …
